### PR TITLE
Handle websoc death

### LINKED ISSUE #532: Users can't see schedule when WebSOC dies
@Voark observed that when WebSOC dies, people can't see their schedules, even though we have it. The UI waits for class details to display. 

We should probably show some raw version of what classes are in each schedule if WebSOC/PeterPortal doesn't respond in time, so that people can still copy the codes and register for classes.

### PR DESCRIPTION
**Pull request by MinhxNguyen7**
## Summary

Backup display for when PP or WebSoc goes down (but our backend is fine).

image

## Test Plan

Block peterportal with devtools. See if it works

## Issues

Closes #532

### CODE REVIEW

**Comment by MinhxNguyen7** [comment]
@Voark I felt a great disturbance in the CI/CD pipeline.

There's no deploy.

**Comment by MinhxNguyen7** [comment]
@ap0nia review pls

**Comment by ap0nia** [comment]
> @ap0nia review pls

ok fine, you do my map ! :pray:

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: I like the semantics of `skeleton`, i.e. `displayScheduleSkeleton`

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: Don't call functions in JSX, this is generally an anti-pattern because they won't be registered by React as components, and so they'll have screwed up state.
**However**, there is a caveat to this rule with class-based components for some reason. But still...don't do it :shrug:

Is this syntax possible at all? Just encapsulate both components entirely as functions, and then render one or the other as a component.

```suggestion
            this.state.barebonesMode ? <this.getBarebonesSchedule /> : <this.getGrid />
```

File: apps/antalmanac/src/stores/AppStore.ts
Comment (ap0nia) [inline apps/antalmanac/src/stores/AppStore.ts]: Just a complaint with the app store and not something you have to fix.
Why is there an abstraction over the schedules? It makes literally no sense. There's one app store in the app, and one schedule store in the app. If you want to subscribe to a particular data source, just pick the one you need.

But coercing the schedules into the schedule store just makes a really bloated class for no reason, with the guise that it's the single source of truth (it isn't and it doesn't make anything simpler).

File: apps/antalmanac/src/stores/Schedules.ts
Comment (ap0nia) [inline apps/antalmanac/src/stores/Schedules.ts]: Here's a funny strat I came up with to load all the possible colors we have lol; it has black and white too tho, so you'd have to deal with them too
https://github.com/icssc/AntAlmanac/blob/aponia-monorepo/apps/client/src/stores/schedule/course.ts#L17

File: apps/antalmanac/src/stores/Schedules.ts
Comment (ap0nia) [inline apps/antalmanac/src/stores/Schedules.ts]: Can we try something like this? I haven' tested it or anything, but I can follow this logic more easily.
```ts
const sectionsGroupedByTen = sectionCodes.reduce((accumulated, item, index) => {
  if (index % 10) {
   accumulated.push(item)
  } else {
    accumulated.push([])
    accumulated[index % 10].push(item)
  }
  return accumulated
}, [] as string[][])

const socObjects = await Promise.all(sectionsGroupedByTen.map(s => { ... }))

const courseInfo = getCourseInfo(combineSOCObjects(socObjects))
```

**Review by ap0nia — CHANGES_REQUESTED** [review]
Mostly looks good, I'm fine with changes to the store stuff because all of it is unfixable anyways.
Just left a suggestion for the component rendering as well as the reducing loop

File: apps/antalmanac/src/stores/AppStore.ts
Comment (MinhxNguyen7) [reply apps/antalmanac/src/stores/AppStore.ts]: I'd imagine that's because schedule used to be a simple data structure that existed in AppStore, then it was extracted when it got too complicated.

Also, if you feel like talking OOP, it's about AppStore owning the schedule class, which is why OOP is not good

File: apps/antalmanac/src/stores/Schedules.ts
Comment (MinhxNguyen7) [reply apps/antalmanac/src/stores/Schedules.ts]: These aren't all MUI colors though. Some of them are too close, and the current ordering specifically avoids close colors for as long as possible.

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (MinhxNguyen7) [reply apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: I like it. I couldn't come up with anything

File: apps/antalmanac/src/stores/AppStore.ts
Comment (ap0nia) [reply apps/antalmanac/src/stores/AppStore.ts]: Based oop hater 😎

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (ap0nia) [reply apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: Oh, actually there's two anti patterns lol.
1. Creating a component inside of another component is the main thing that screws up state (unless it's a class component for some reason). Correct practice is to make a standalone component...even if it requires prop-drilling.

2. Using a function call instead of a React JSX tag thingy. This just prevents it from rendering on the component tree properly or something.

Conclusion : I love how React is so **easy to use** ™️

File: apps/antalmanac/src/stores/Schedules.ts
Comment (ap0nia) [inline apps/antalmanac/src/stores/Schedules.ts]: Ok, I did some testing and I think this should work.
```suggestion
                    const sectionsGroupedByTen = sectionCodes.reduce((accumulated, item, index) => {
                        if (index % 10) {
                            accumulated.at(-1)?.push(item);
                        } else {
                            accumulated.push([item]);
                        }
                        return accumulated;
                    }, [] as string[][]);
```

File: apps/antalmanac/src/components/PatchNotes.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/PatchNotes.tsx:1]: This component was written really poorly, so I refactored it real quick.

File: apps/antalmanac/src/components/Calendar/Toolbar/ExportCalendar.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/Calendar/Toolbar/ExportCalendar.tsx:1]: This file was written poorly, so I refactored it.

File: apps/antalmanac/tsconfig.json
Comment (ap0nia) [inline apps/antalmanac/tsconfig.json]: Extra commas is invalid in JSON; TypeScript doesn't mind when parsing the file, but following the spec is good.

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx:1]: Ok, I just realized the UI of this component was implemented completely wrong and I fixed it.

**Review by ap0nia — APPROVED** [review]
Ok, I took a crack at correcting the discrepancies between this branch and main. Let me know if you think it's good and we can move forward with merging.

**Comment by MinhxNguyen7** [comment]
Patch notes now keep showing up on reload. I can fix that really quickly

Otherwise, I don't see anything amiss, other than that this is a _cosmic commit_

**Review by ap0nia — APPROVED** [review]
I reviewed all of the stuff while I was refactoring it, and I'm fine with how it currently is lol. If you'd like anything adjusted, let me know or you can apply changes yourself. :nerd_face:

**Comment by MinhxNguyen7** [comment]
Well I just noticed that the central feature of this PR actually doesn't work lol. It still does the same thing as `main`. I think the TRPC PR changed how schedules are loaded. I'll try to fix that now

File: apps/antalmanac/package.json
Comment (ap0nia) [inline apps/antalmanac/package.json:72]: I've decided that I'm ok with the pnpm default of installing workspace packages with the semantic version `workspace:^`

File: apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/Calendar/CourseCalendarEvent.tsx:158]: Technically, only stateful variables are needed in the dependency array, i.e. things that could change or be mutated. State setters should never change, but this is React and keeping track of what things change is stupid.

In previous commits, I did add the state setter to the dependency array to be safe. I'm gonna chalk this up to LGTM :nerd:

File: apps/antalmanac/src/components/RightPane/RightPaneRoot.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/RightPane/RightPaneRoot.tsx]: In general, we should try to avoid introducing specific edge cases like this. i.e. this particular div should render everything with the exact same padding, and then have the particular component position itself -- i.e. since it knows how it should be rendered.

I did address this in some other PR, so this is fine to leave as-is.

File: apps/antalmanac/src/stores/AppStore.ts
Comment (ap0nia) [inline apps/antalmanac/src/stores/AppStore.ts:19]: In general, I like to use `Record` if possible because it's more succinct. A niche edge case where this isn't very nice is when you want to map an interface's properties to alternate values. A `Record` isn't able to capture the original definition of the properties.

[Here's an example on TS playground](https://www.typescriptlang.org/play?ssl=3&ssc=12&pln=2&pc=12#code/C4TwDgpgBAsiCSA7YEBOAzAhgY2gXigG8AoKKTALigGdhUBLRAc1KgCMraHnXtO7GLAL7FioSLBAAlCNgD2qACZQCM+UoA8Aawgg56SUhQYcEADQ0BzAHyiANhGBQAHlThqFyggFdEiiOiMEMqY1OSIIKIA9FFQAKrUglAARExyUMDp-oGI9MD0cojJUIUZABbQyZjFYKhykKigUIpyENSIAOROTI5Qet7hIADuFagQxDHsspje1NDAFVAABh5KSxng0ABWs07oCj3AYSNo84sK9EyMmHZQtfVo+W1Q2JgAttDodW8Tsb7+qHK0DKcjkigseQ6YTY01m0DyUB2tBehS43mwR3IUF8BUQJQMN1uC2gOhA1AhiFoEEwyn05Ax3kJICgb0wYDASS+cjeQJKDCuiBuUFJxGcADpMKJxNA4DA2ZBFEY0FhcCoiFAANoAaSgjGFujpcCVJlwAF1+NwmFARMQHE4QG4QHL2cFjSr8Ni-AEgiEwpgItFYgkkql0plmt7cvlCsVSsSUtU7nUGk1gJgdGF+uwcFoMukjchlaYNR1MB1Tb8pq84VjWS7lGN1A2IB83jDUGF4xcBUL7inmQihqEIwwAG7BKBcn4gCVAA)

In GitHub markdown:
```ts
type MyInterface = {
  a: string
  b: string
  c: string
}

type MyRecord = Record<keyof MyInterface, string>

let x: MyRecord = undefined as any

// Using "go to definition" on the "a" property doesn't get you anywhere
// because the `Record` type just forgets where the original properties came from
// under the hood, it's because it just constructs a union of all the keys, instead of actually mapping from the original key
x.a

type MyMappedInterface = { [K in keyof MyInterface]: string }

let y: MyMappedInterface = undefined as any

// Using "go to definition" on the "a" property takes you back to MyInterface['a']
// because a mapped record remembers the original property it was derived from
y.a
```

**Review by ap0nia — APPROVED** [review]
I think these changes look solid, good work! 🙌 
Let me know if you're ok with finally merging everything! @MinhxNguyen7

File: apps/antalmanac/src/components/RightPane/RightPaneRoot.tsx
Comment (MinhxNguyen7) [reply apps/antalmanac/src/components/RightPane/RightPaneRoot.tsx]: Yeah I thought it was janky but out of scope to change in this PR

File: apps/antalmanac/src/stores/AppStore.ts
Comment (MinhxNguyen7) [reply apps/antalmanac/src/stores/AppStore.ts:19]: I don't disagree, but I'd argue that it's out of scope. I'll leave it since you've already done it.

**Comment by MinhxNguyen7** [comment]
I think the PR's good now, but I can't review it because I'm the author. @ap0nia, you shouldn't review it because you wrote a lot of the code here. We just need someone to look through and check that things are sound and that we haven't broken anything.

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx:225]: Color should probably be `inherit` like on main

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]:
```suggestion
                        color="inherit"
```
Currently invisible on light mode

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]:
```suggestion
                                <Button variant="outlined" color="inherit" {...bindTrigger(popupState)}>
```
Currently it's blue colored, but it should match what's on live

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: Margins are also a little weird around this title section:

Screenshot 2023-09-06 at 3 01 11 AM

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx:272]: This is not updating when custom events are created:

chrome-capture-2023-8-6

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: This is not updating when custom events are created:

chrome-capture-2023-8-6

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: Styling is off here:

Screenshot 2023-09-06 at 3 04 36 AM

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: Sorry about all the styling nitpicks, but can we add a little padding here:

Screenshot 2023-09-06 at 1 39 25 PM

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]:
```suggestion
        <>
                        <Typography variant="h6">{`${skeletonSchedule.scheduleName}`}</Typography>

```

Looking at `ShortCourseScheduleSchema`, we should be able to improve the user experience by also supporting displaying the name of the schedule. 

_You may need to restyle the term typography a little bit, depending on how you decide to implement the typography of the schedule name._

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (ap0nia) [reply apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: Oh, I meant to wrap a padded box around the entire thing, but I forgot.

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]:
```suggestion
                                allowHighlight={false}
```

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [reply apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: In a similar vein, I think it... should be possible to display custom events and schedule notes as well?

The base functionality is great and I think adding on this additional support would really make it awesome

Comment (ap0nia) [reply apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: Can we add that later lol; I don't wanna debug more features

Comment (KevinWu098) [reply apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: Sure, let's hold off on custom events and notes. But the one thing I will really really push for right now is having the schedule name shown.

Otherwise, there's no real way for users to figure out what's what based off just the codes.

Comment (ap0nia) [reply apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: Did I not do that? https://github.com/icssc/AntAlmanac/blob/de7f86b5ca63b22e5bdc1f952144448ccfe2a826/apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx#L110

Comment (KevinWu098) [reply apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]: 💀 sorry, I'm on mobile and didn't see it 😅

I'll check it out once I'm home; Thanks for adding in the schedule name functionality :))

File: apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneRoot.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneRoot.tsx]: Not sure where the source of the issue is, but the GIF has been shifted up in the Search pane:

Screenshot 2023-09-06 at 8 07 06 PM

**Comment by KevinWu098** [comment]
```typescript
export const ShortCourseScheduleSchema = type([
    {
        scheduleName: 'string',
        courses: arrayOf(ShortCourseSchema),
        customEvents: arrayOf(RepeatingCustomEventSchema),
        'scheduleNote?': 'string',
    },
    '|>',
    (s) => ({ scheduleNote: '', ...s }),
]);
export type ShortCourseSchedule = typeof ShortCourseScheduleSchema.infer;
```

~~I'm noticing that Typescript can't read the type of `ShortCourseSchedule`. My guess is because of the syntax used for scheduleNote? Is there a way for us to nudge Typescript into inferring properly?~~

Screenshot 2023-09-06 at 8 16 52 PM

~~This currently affects `Schedules.ts` and `AddedCoursePane.tsx`~~

Nvm I'm trolling, fixing this is out of scope I think, but we should probably fix the types in a soon <sup>tm</sup> PR since this has been antagonizing `Schedules.ts` for some time 😅

**Comment by KevinWu098** [comment]
Honestly, I really love this support for when Websoc dies, and I think future follow-up could make it an even better user experience.

For now, the comments above are my only remaining issues (I hope), but just to list out some future targets:
1. Support custom events and schedule notes
2. Disable the calendar toolbar (since you can't do anything with it anyways)
3. Support toggling between multiple schedules in skeleton mode

File: apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneRoot.tsx
Comment (ap0nia) [reply apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneRoot.tsx]: Is it supposed to be in the middle?

**Comment by ap0nia** [comment]
The type issues should a problem on your end; not a universal one. You could try asking for help, opening an issue to start a discussion, etc. but strictly speaking this isn't "solvable", since everything works fine on my machine :tm:

File: apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneRoot.tsx
Comment (KevinWu098) [reply apps/antalmanac/src/components/RightPane/CoursePane/CoursePaneRoot.tsx]: Yeah, center of the pane. You can see the intended behavior on the Map pane

**Review by KevinWu098 — APPROVED** [review]
I would nitpick the styling here, but I think it's not worth holding this PR up over; It'll get obliterated anyways after this gets merged in and I add the intended changes from #665 

Screenshot 2023-09-07 at 4 06 45 AM

LGTM, awesome feature 💖

I'll leave the merging to y'alls, just in case anybody wants to double-check my review 🫶

**Comment by KevinWu098** [comment]
> The type issues should a problem on your end; not a universal one. You could try asking for help, opening an issue to start a discussion, etc. but strictly speaking this isn't "solvable", since everything works fine on my machine ™️

💀 I'll figure it out one of these days

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]:
```suggestion
import { useCallback, useEffect, useMemo, useState } from 'react';
```

Unused import

File: apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/RightPane/AddedCourses/AddedCoursePane.tsx]:
```suggestion
```
Unused import

**Review by KevinWu098 — APPROVED** [review]
LGTM 🚀

**Pull request merged**
